offsetresolve.go

```diff
@@ -0,0 +1,438 @@
+			return nil, err
+		}
+	}
+	return resolved, nil
+}
+
+// expandAllIntervalsSpec expands the "all partitions" entry in offsets if present,
+// creating entries for all partitions not explicitly specified in offsets.
+//
+// If there isn't enough information in q to do that, it leaves
+// the offsets map unchanged.
+func (r *resolver) expandAllIntervalsSpec(offsets map[int32]*interval, q *offsetQuery) {
+	intv, ok := offsets[-1]
+	if !ok {
+		// No "all partitions" entry.
+		return
+	}
+	// updateTimestamp updates the "summary" timestamp *t according to p.
+	// It only affects t if p has either an "oldest" or "newest" anchor with
+	// a duration-based difference which implies that we need to find either the
+	// oldest or newest timestamp across all partitions so that the
+	// interval end time is the same across all partitions.
+	updateTimestamp := func(t *time.Time, partition int32, p position) bool {
+		if p.anchor.offset >= 0 || !p.diff.isDuration || p.anchor.offset == offsetResume {
+			// We can work out the "all" offset for each partition independently.
+			// Lacking a better idea, we leave resume as per-partition.
+			return true
+		}
+		off, ok := r.info.getOffset(partition, symbolicOffsetRequest(p.anchor.offset), q)
+		if !ok {
+			return false
+		}
+		t1, ok := r.info.getTime(partition, off, q)
+		if !ok {
+			return false
+		}
+		if t1.IsZero() {
+			// No timestamp available (probably because the partition is empty)
+			return true
+		}
+		switch p.anchor.offset {
+		case sarama.OffsetOldest:
+			if t1.Before(*t) {
+				*t = t1
+			}
+		case sarama.OffsetNewest:
+			if (*t).IsZero() || t1.After(*t) {
+				*t = t1
+			}
+		}
+		return true
+	}
+	var startTime, endTime time.Time
+	gotAll := true
+	for _, partition := range r.allPartitions {
+		if _, ok := offsets[partition]; ok {
+			// The partition was explicitly specified, so it's independent.
+			continue
+		}
+		gotAll = updateTimestamp(&startTime, partition, intv.start) && gotAll
+		gotAll = updateTimestamp(&endTime, partition, intv.end) && gotAll
+	}
+	if !gotAll {
+		return
+	}
+	// We've got enough information to fill out the offsets we need.
+	// First update the "all partitions" interval to be the correct time if needed.
+	// Note that if startTime or endTime are zero (because all partitions
+	// are empty for example), we'll leave the interval to be resolved later,
+	// which should work out fine in the end.
+	if !startTime.IsZero() {
+		intv.start.anchor.isTime = true
+		intv.start.anchor.time = startTime.Add(intv.start.diff.duration)
+		intv.start.diff = anchorDiff{}
+	}
+	if !endTime.IsZero() {
+		intv.end.anchor.isTime = true
+		intv.end.anchor.time = endTime.Add(intv.end.diff.duration)
+		intv.end.diff = anchorDiff{}
+	}
+	delete(offsets, -1)
+	for _, partition := range r.allPartitions {
+		if _, ok := offsets[partition]; ok {
+			continue
+		}
+		intv := *intv
+		offsets[partition] = &intv
+	}
+}
+
+func min(x, y int64) int64 {
```

Comment:
Mildly annoying, isn't it, that the standard library only provides this for float64.